OSDOCS-11615: TCP flags filtering + SYN flooding alerts and metrics (#80610)
openshift/openshift-docs: modules/network-observability-tcp-flag-syn-flood.adoc

. In the *Provided APIs* heading for the *NetObserv Operator*, select *FlowMetric*.
. In the *Project:*  dropdown list, select the project of the Network Observability Operator instance. 
. Click *Create FlowMetric*.
. Create `FlowMetric` resources to add the following configurations: 
+
.FlowMetric configuration to count flows per destination host and resource, with TCP flags
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-destination
spec:
  metricName: flows_with_flags_per_destination_total
  type: Counter
  labels: [SrcSubnetLabel,DstSubnetLabel,DstK8S_Name,DstK8S_Type,DstK8S_HostName,DstK8S_Namespace,Flags]
----
====
+
.FlowMetric configuration to count flows per source host and resource, with TCP flags
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-source
spec:
  metricName: flows_with_flags_per_source_total
  type: Counter
  labels: [DstSubnetLabel,SrcSubnetLabel,SrcK8S_Name,SrcK8S_Type,SrcK8S_HostName,SrcK8S_Namespace,Flags]
----
====

> 

Reviewer: there's something missing here, the whole rest of the file looks like YAML

Author: Ok cool 😅

          {{ $labels.job }}: incoming SYN-flood attack suspected to Host={{ $labels.DstK8S_HostName}}, Namespace={{ $labels.DstK8S_Namespace }}, Resource={{ $labels.DstK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Incoming SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_destination_total{Flags="2"}[1m])) by (job, DstK8S_HostName, DstK8S_Namespace, DstK8S_Name) > 300      
      for: 15s
      labels:
        severity: warning
        app: netobserv
    - alert: NetObserv-SYNFlood-out
      annotations:
        message: |-
          {{ $labels.job }}: outgoing SYN-flood attack suspected from Host={{ $labels.SrcK8S_HostName}}, Namespace={{ $labels.SrcK8S_Namespace }}, Resource={{ $labels.SrcK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Outgoing SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_source_total{Flags="2"}[1m])) by (job, SrcK8S_HostName, SrcK8S_Namespace, SrcK8S_Name) > 300      
      for: 15s
      labels:
        severity: warning
        app: netobserv
# ...
----

> . Verify that in the OpenShift web console *Network Traffic* table view, click *Manage Columns* -> *TCP flags*. You can observe the *TCP flags* column.

Author: @Amoghrd I added this piece but I'm unsure if it sorta implies that you must do all these preceding steps before you can see the TCP Flags column in the Network Traffic table. Maybe its worth just nothing that the TCP flags column is available to be enabled even without these preceding steps. I also added a small heading in the Observing network traffic section to kinda tie back to this: https://80610--ocpdocs-pr.netlify.app/openshift-enterprise/latest/observability/network_observability/observing-network-traffic#network-observability-filtering-tcp-flags_nw-observe-network-traffic

        summary: "Incoming SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_destination_total{Flags="2"}[1m])) by (job, DstK8S_HostName, DstK8S_Namespace, DstK8S_Name) > 300      
      for: 15s
      labels:
        severity: warning
        app: netobserv
    - alert: NetObserv-SYNFlood-out
      annotations:
        message: |-
          {{ $labels.job }}: outgoing SYN-flood attack suspected from Host={{ $labels.SrcK8S_HostName}}, Namespace={{ $labels.SrcK8S_Namespace }}, Resource={{ $labels.SrcK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Outgoing SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_source_total{Flags="2"}[1m])) by (job, SrcK8S_HostName, SrcK8S_Namespace, SrcK8S_Name) > 300      
      for: 15s
      labels:
        severity: warning
        app: netobserv
# ...
----
.Verification
. In the web console, click *Manage Columns* in the *Network Traffic* table view and click *TCP flags*.
. In the *Network Traffic* table view, filter on *TCP protocol SYN TCPFlag*. Large number of flows with same *byteSize* indicate a SYN flood.

> . Go to *Observe* -> *Alerting* -> *Alerting Rules* tab and filter on *netobserv-synflood-in alert*. An alert displays when SYN flooding occurs. 

Reviewer 2: I think we should add something about the threshold that is used in the alerts (300) : users should adapt it empirically. A threshold that is too low may produce false-positives, and if it's too high it may miss actual attacks. It is to be determined empirically.

. In the *Provided APIs* heading for the *NetObserv Operator*, select *FlowMetric*.
. In the *Project:*  dropdown list, select the project of the Network Observability Operator instance. 
. Click *Create FlowMetric*.
. Create `FlowMetric` resources to add the following configurations: 
+
.FlowMetric configuration to count flows per destination host and resource, with TCP flags
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-destination
spec:
  metricName: flows_with_flags_per_destination_total
  type: Counter
  labels: [SrcSubnetLabel,DstSubnetLabel,DstK8S_Name,DstK8S_Type,DstK8S_HostName,DstK8S_Namespace,Flags]
----
====
+
.FlowMetric configuration to count flows per source host and resource, with TCP flags
[%collapsible]
====
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-source
spec:
  metricName: flows_with_flags_per_source_total
  type: Counter
  labels: [DstSubnetLabel,SrcSubnetLabel,SrcK8S_Name,SrcK8S_Type,SrcK8S_HostName,SrcK8S_Namespace,Flags]
----
====

> . Deploy the following `AlertingRule` to alert for SYN flooding:

Reviewer 3: Is there a term that can be added after `AlertingRule` for clarity? Like resource?
```suggestion
. Deploy the following `AlertingRule` resource to alert for SYN flooding:
```

Author: Good call

+
.Alert for SYN flooding
[source,yaml]
----
apiVersion: monitoring.openshift.io/v1
kind: AlertingRule
metadata:
  name: netobserv-syn-alerts
  namespace: openshift-monitoring
# ...
  spec:
  groups:
  - name: NetObservSYNAlerts
    rules:
    - alert: NetObserv-SYNFlood-in
      annotations:
        message: |-
          {{ $labels.job }}: incoming SYN-flood attack suspected to Host={{ $labels.DstK8S_HostName}}, Namespace={{ $labels.DstK8S_Namespace }}, Resource={{ $labels.DstK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Incoming SYN-flood"

>       expr: sum(rate(netobserv_flows_with_flags_per_destination_total{Flags="2"}[1m])) by (job, DstK8S_HostName, DstK8S_Namespace, DstK8S_Name) > 300      

Reviewer: the `<1>` should be added here as well

      for: 15s
      labels:
        severity: warning
        app: netobserv
    - alert: NetObserv-SYNFlood-out
      annotations:
        message: |-
          {{ $labels.job }}: outgoing SYN-flood attack suspected from Host={{ $labels.SrcK8S_HostName}}, Namespace={{ $labels.SrcK8S_Namespace }}, Resource={{ $labels.SrcK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Outgoing SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_source_total{Flags="2"}[1m])) by (job, SrcK8S_HostName, SrcK8S_Namespace, SrcK8S_Name) > 300       <1>
      for: 15s
      labels:
        severity: warning
        app: netobserv
# ...
----
<1> In this example, the threshold for the alert is `300`; however, you can adapt this value empirically. A threshold that is too low might produce false-positives, and if it's too high it might miss actual attacks. It is to be determined empirically.

.Verification
. In the web console, click *Manage Columns* in the *Network Traffic* table view and click *TCP flags*.

> . In the *Network Traffic* table view, filter on *TCP protocol SYN TCPFlag*. Large number of flows with same *byteSize* indicate a SYN flood.

Reviewer 3: This might sound better with an article at the beginning:
```suggestion
. In the *Network Traffic* table view, filter on *TCP protocol SYN TCPFlag*. A large number of flows with same *byteSize* indicate a SYN flood.
```

> . In the *Network Traffic* table view, filter on *TCP protocol SYN TCPFlag*. Large number of flows with same *byteSize* indicate a SYN flood.

Reviewer 3: Idk if I'm thinking about this correctly but if number is singular, shouldn't indicate have an s at the end? Since it's the number that's doing the indication
```suggestion
. In the *Network Traffic* table view, filter on *TCP protocol SYN TCPFlag*. Large number of flows with same *byteSize* indicates a SYN flood.
```

Author: yeah I think I'm missing a couple of articles and also this verb isn't correct. Good points!

> . Go to *Observe* -> *Alerting* -> *Alerting Rules* tab and filter on *netobserv-synflood-in alert*. An alert displays when SYN flooding occurs. 

Reviewer: Last line should be updated to: `The alert should be firing when SYN flooding occurs`

Reviewer: this change has not been incorporated yet

> . Go to *Observe* -> *Alerting* -> *Alerting Rules* tab and filter on *netobserv-synflood-in alert*. The alert should fire when SYN flooding occurs. 

Reviewer 3: I feel like the GUI direction here is a bit off, and I can't tell what the right answer would be without the GUI in front of me. Is it that you go to the Observe page, then the Alerting page, and then select the Alerting Rules tab? Or are all three of those different tabs? 

I'm not sure what the best answer would be here, but either removing tab or breaking up that set of three GUI stops might be the accurate description:
```suggestion
. Go to *Observe* -> *Alerting* -> *Alerting Rules* and filter on *netobserv-synflood-in alert*. The alert should fire when SYN flooding occurs. 
```

```suggestion
. Go to *Observe* -> *Alerting* and select the *Alerting Rules* tab, then filter on *netobserv-synflood-in alert*. The alert should fire when SYN flooding occurs. 
```

Reviewer 2: I think the second suggestion makes sense. "Observe > Alerting" is just a sub-menu item to navigate to. "Alerting Rules" is a tab to click on.

> . Go to *Observe* -> *Alerting* -> *Alerting Rules* tab and filter on *netobserv-synflood-in alert*. The alert should trigger when SYN flooding occurs. 

Author: @Amoghrd I updated this to "The alert should trigger when SYN flooding occurs. " Is this where you meant for me to put it?

Reviewer: It should be `firing` instead of `trigger`

Author: I saw both `trigger` and `fire` used in the Monitoring documentation as the action that happens "when conditions are true", and I was thinking `trigger` is less ambiguous for translation and global audiences than `fire` so I used `trigger`. But I can use `fire`. Seems like both are used in the Monitoring documentation anyways. 

https://docs.openshift.com/container-platform/4.16/observability/monitoring/managing-alerts.html

Reviewer: oh I did know about the monitoring doc nomenclature. We see Firing on console when the alert is hit so wanted to keep it the same way

// Module included in the following assemblies:
//
// network_observability/metrics-alerts-dashboards.adoc

:_mod-docs-content-type: PROCEDURE
[id="network-observability-tcp-flag-syn-flood_{context}"]
= Detecting SYN flooding using the FlowMetric API and TCP flags

> You can create an `alertRule` to alert for SYN flooding. 

Reviewer 4: This says "`alertRule`" but later we have "`AlertingRule` resource". Should it be one of the following?

```suggestion
You can create an alert rule to alert for SYN flooding. 
```

```suggestion
You can create an `AlertingRule` resource to alert for SYN flooding. 
```

.Procedure
. In the web console, navigate to *Operators* -> *Installed Operators*.
. In the *Provided APIs* heading for the *NetObserv Operator*, select *FlowMetric*.

> . In the *Project:*  dropdown list, select the project of the Network Observability Operator instance. 

Reviewer 4: Generally should avoid including punctuation in UI labels
```suggestion
. In the *Project*  dropdown list, select the project of the Network Observability Operator instance. 
```

. Click *Create FlowMetric*.

> . Create `FlowMetric` resources to add the following configurations: 
> +
> .FlowMetric configuration to count flows per destination host and resource, with TCP flags
> [%collapsible]
> ====
> [source,yaml]
> ----
> apiVersion: flows.netobserv.io/v1alpha1
> kind: FlowMetric
> metadata:
>   name: flows-with-flags-per-destination
> spec:
>   metricName: flows_with_flags_per_destination_total
>   type: Counter
>   labels: [SrcSubnetLabel,DstSubnetLabel,DstK8S_Name,DstK8S_Type,DstK8S_HostName,DstK8S_Namespace,Flags]
> ----
> ====
> +
> .FlowMetric configuration to count flows per source host and resource, with TCP flags
> [%collapsible]
> ====
> [source,yaml]
> ----
> apiVersion: flows.netobserv.io/v1alpha1
> kind: FlowMetric
> metadata:
>   name: flows-with-flags-per-source
> spec:
>   metricName: flows_with_flags_per_source_total
>   type: Counter
>   labels: [DstSubnetLabel,SrcSubnetLabel,SrcK8S_Name,SrcK8S_Type,SrcK8S_HostName,SrcK8S_Namespace,Flags]
> ----
> ====

Reviewer 4: Collapsible blocks are not supported on docs.redhat.com :crying_cat_face: Should do these as standard YAML samples, might want to shorten the titles a bit as a result:

```suggestion
. Create `FlowMetric` resources to add the following configurations: 
+
.Configuration counting flows per destination host and resource, with TCP flags
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-destination
spec:
  metricName: flows_with_flags_per_destination_total
  type: Counter
  labels: [SrcSubnetLabel,DstSubnetLabel,DstK8S_Name,DstK8S_Type,DstK8S_HostName,DstK8S_Namespace,Flags]
----
+
.Configuration counting flows per source host and resource, with TCP flags
[source,yaml]
----
apiVersion: flows.netobserv.io/v1alpha1
kind: FlowMetric
metadata:
  name: flows-with-flags-per-source
spec:
  metricName: flows_with_flags_per_source_total
  type: Counter
  labels: [DstSubnetLabel,SrcSubnetLabel,SrcK8S_Name,SrcK8S_Type,SrcK8S_HostName,SrcK8S_Namespace,Flags]
----
```

Author: Sad but heard

. Deploy the following `AlertingRule` resource to alert for SYN flooding:
+

> .Alert for SYN flooding

Author: changing this to `AlertingRule` for SYN flooding

          {{ $labels.job }}: incoming SYN-flood attack suspected to Host={{ $labels.DstK8S_HostName}}, Namespace={{ $labels.DstK8S_Namespace }}, Resource={{ $labels.DstK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Incoming SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_destination_total{Flags="2"}[1m])) by (job, DstK8S_HostName, DstK8S_Namespace, DstK8S_Name) > 300      <1>
      for: 15s
      labels:
        severity: warning
        app: netobserv
    - alert: NetObserv-SYNFlood-out
      annotations:
        message: |-
          {{ $labels.job }}: outgoing SYN-flood attack suspected from Host={{ $labels.SrcK8S_HostName}}, Namespace={{ $labels.SrcK8S_Namespace }}, Resource={{ $labels.SrcK8S_Name }}. This is characterized by a high volume of SYN-only flows with different source IPs and/or ports.
        summary: "Outgoing SYN-flood"
      expr: sum(rate(netobserv_flows_with_flags_per_source_total{Flags="2"}[1m])) by (job, SrcK8S_HostName, SrcK8S_Namespace, SrcK8S_Name) > 300       <1>
      for: 15s
      labels:
        severity: warning
        app: netobserv
# ...
----

> <1> In this example, the threshold for the alert is `300`; however, you can adapt this value empirically. A threshold that is too low might produce false-positives, and if it's too high it might miss actual attacks. It is to be determined empirically.

Reviewer 4: Seems like a repeat? If this needs to stay, no worries :)
```suggestion
<1> In this example, the threshold for the alert is `300`; however, you can adapt this value empirically. A threshold that is too low might produce false-positives, and if it's too high it might miss actual attacks.
```
